### PR TITLE
Re-architected server query execution to hand over query to the scheduler for execution.

### PR DESCRIPTION
**Pull request by antumbde**
This commit changes the server to pass query request to the
scheduler for execution instead of executing the query in
the receiving Netty server thread. Scheduler by default has two
executor pools - QueryRunners which will execute the query, and
QueryWorkers which are used by the query runtime to parallelize
operators across segments. Added a First Come First Served query
scheduler that matches existing scheduling functionality by
immediately sending the query to the QueryRunners for execution.
This commit also adds a ScheduledRequestHandler which _will_ replace
SimpleRequestHandler to use QueryScheduler in future.

Current code continues to use SimpleRequestHandler. The impacting change
of this commit is to run queries in separate executors and to return
ListenableFutures to the caller.

Testing:
1. Additional unit tests and existing integration tests
2. Query execution on local cluster
3. Performance tests

### CODE REVIEW

File: pinot-transport/src/test/java/com/linkedin/pinot/transport/netty/NettyCloseChannelTest.java [hunk 1]
```diff
@@ -174,7 +176,7 @@ protected void initChannel(SocketChannel ch) throws Exception {
       super.initChannel(ch);
       ch.pipeline().removeLast();
       ch.pipeline().addLast("request_handler", new MyNettyChannelInboundHandler());
-      LOGGER.info("Server Channel pipeline setup modified. Pipeline:" + ch.pipeline().names());
+      LOGGER.debug("Server Channel pipeline setup modified. Pipeline:" + ch.pipeline().names());
```
Comment (jfim) [inline pinot-transport/src/test/java/com/linkedin/pinot/transport/netty/NettyCloseChannelTest.java]: Remove string concatenation

File: pinot-server/src/main/java/com/linkedin/pinot/server/conf/ServerConf.java [hunk 2]
```diff
@@ -78,8 +83,16 @@ public String getQueryExecutorClassName() {
...
+    return _serverConf.getString(PINOT_SERVER_QUERY_SCHEDULER_CLASS_CONFIG_KEY,
+        DEFAULT_QUERY_SCHEDULER_CLASS_NAME);
+  }
+
   public String getRequestHandlerFactoryClassName() {
     return _serverConf.getString(PINOT_SERVER_REQUEST_HANDLER_FACTORY_CLASS);
   }
 
+  public Configuration getSchedulerConfig() {
+    return _serverConf.subset(PINOT_ + ".query.scheduler");
```
Comment (jfim) [inline pinot-server/src/main/java/com/linkedin/pinot/server/conf/ServerConf.java]: Use PINOT_QUERY_SCHEDULER_PREFIX

File: pinot-server/src/test/java/com/linkedin/pinot/server/request/ScheduledRequestHandlerTest.java [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+    serverMetrics = new ServerMetrics(new MetricsRegistry());
+    channelHandlerContext = mock(ChannelHandlerContext.class);
+    queryScheduler = mock(QueryScheduler.class);
+  }
+
+  @Test
+  public void testBadRequest()
+      throws Exception {
+    ScheduledRequestHandler handler = new ScheduledRequestHandler(queryScheduler, serverMetrics);
+    String requesBadString = "foobar";
```
Comment (jfim) [inline pinot-server/src/test/java/com/linkedin/pinot/server/request/ScheduledRequestHandlerTest.java]: Typo in request

File: pinot-transport/src/test/java/com/linkedin/pinot/transport/netty/NettyCloseChannelTest.java (hunk 1 above)
Comment (antumbde) [inline pinot-transport/src/test/java/com/linkedin/pinot/transport/netty/NettyCloseChannelTest.java]: Fixed this one but these are existing statements in test code. Not doing full cleanup in this PR

File: pinot-core/src/main/java/com/linkedin/pinot/core/query/scheduler/FCFSQueryScheduler.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+  public FCFSQueryScheduler(@Nonnull Configuration schedulerConfig, @Nonnull QueryExecutor queryExecutor) {
+    super(schedulerConfig, queryExecutor);
+    Preconditions.checkNotNull(queryExecutor);
+  }
+
+  @Override
+  public ListenableFuture<DataTable> submit(final QueryRequest queryRequest) {
+    // Executor will maintain the queue so directly pass on the query to executor
+    // I think submit is thread-safe
```
Comment (mcvsubbu) [inline pinot-core/src/main/java/com/linkedin/pinot/core/query/scheduler/FCFSQueryScheduler.java]: Can you remove this comment (after verifying it is thread-safe, of course :-)

File: pinot-server/src/main/java/com/linkedin/pinot/server/request/ScheduledRequestHandler.java [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+    try {
+      if (instanceResponse == null) {
+        LOGGER.warn("Instance response is null.");
+        responseByte = new byte[0];
+      } else {
+        responseByte = instanceResponse.toBytes();
+      }
+    } catch (Exception e) {
+      serverMetrics.addMeteredGlobalValue(ServerMeter.RESPONSE_SERIALIZATION_EXCEPTIONS, 1);
+      LOGGER.error("Got exception while serializing response.", e);
```
Comment (mcvsubbu) [inline pinot-server/src/main/java/com/linkedin/pinot/server/request/ScheduledRequestHandler.java]: add request ID here?

File: pinot-server/src/main/java/com/linkedin/pinot/server/request/ScheduledRequestHandler.java [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+        if (instanceResponse == null) {
+          LOGGER.warn("Instance response is null for requestId: {}", instanceRequest.getRequestId());
+          return new byte[0];
+        }
+        byte[] responseBytes;
+        try {
+          responseBytes = instanceResponse.toBytes();
+        } catch (Exception e) {
+          serverMetrics.addMeteredGlobalValue(ServerMeter.RESPONSE_SERIALIZATION_EXCEPTIONS, 1);
+          LOGGER.error("Got exception while serializing response.", e);
```
Comment (mcvsubbu) [inline pinot-server/src/main/java/com/linkedin/pinot/server/request/ScheduledRequestHandler.java]: add request ID

Comment (mcvsubbu) [inline pinot-server/src/main/java/com/linkedin/pinot/server/request/ScheduledRequestHandler.java]: and broker ID

File: pinot-server/src/main/java/com/linkedin/pinot/server/request/ScheduledRequestHandler.java (hunk 5 above)
Comment (mcvsubbu) [inline pinot-server/src/main/java/com/linkedin/pinot/server/request/ScheduledRequestHandler.java]: and brokeriD

File: pinot-server/src/main/java/com/linkedin/pinot/server/request/ScheduledRequestHandler.java [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+    // transform the DataTable to serialized byte[] to send back to broker
+    ListenableFuture<byte[]> serializedQueryResponse = Futures.transform(queryResponse, new Function<DataTable, byte[]>() {
+      @Nullable
+      @Override
+      public byte[] apply(@Nullable DataTable instanceResponse) {
+        long totalNanos = System.nanoTime() - schedulerSubmitTime;
+        long serializationStartTime = System.nanoTime();
+        serverMetrics.addPhaseTiming(brokerRequest, ServerQueryPhase.QUERY_PROCESSING, totalNanos);
+        if (instanceResponse == null) {
+          LOGGER.warn("Instance response is null for requestId: {}", instanceRequest.getRequestId());
```
Comment (mcvsubbu) [inline pinot-server/src/main/java/com/linkedin/pinot/server/request/ScheduledRequestHandler.java]: good to add broker id in the log

File: pinot-transport/src/main/java/com/linkedin/pinot/transport/netty/NettyServer.java [hunk 8]
```diff
@@ -193,69 +191,74 @@ public NettyChannelInboundHandler(RequestHandler handler, NettyServerMetrics met
...
+            @Override
+            public void operationComplete(ChannelFuture future)
+                throws Exception {
+              LOGGER.debug("Response has been sent !!");
+              responseSendLatency.stop();
+              _metric.addServingStats(requestSizeInBytes, result.length, 1L, false,
+                  requestProcessingLatency.getLatencyMs(), responseSendLatency.getLatencyMs());
+              long totalQueryTime = System.currentTimeMillis() - requestStartTime;
+              if (totalQueryTime > _defaultLargeQueryLatencyMs) {
+                LOGGER.info("Trace Info: request handler processing time : {}, send response latency: {}, total time to handle request: {}",
```
Comment (mcvsubbu) [inline pinot-transport/src/main/java/com/linkedin/pinot/transport/netty/NettyServer.java]: Now is a good time to change "Trace Info" to "Slow query" or something like that.

**Comment by mcvsubbu** [comment]
It is good to have the request Id and Broker id on all logs after the request is decoded.

File: pinot-core/src/main/java/com/linkedin/pinot/core/query/scheduler/QueryScheduler.java [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+  // like poolPerTable or QoS based pools etc
+  // The policy should also determine how many threads we use per query
+  protected ListeningExecutorService queryWorkers;
+
+  final QueryExecutor queryExecutor;
+  static {
+    int numCores = Runtime.getRuntime().availableProcessors();
+    // arbitrary...but not completely arbitrary
+    DEFAULT_QUERY_RUNNER_THREADS = numCores;
+    DEFAULT_QUERY_WORKER_THREADS = 2 * numCores;
```
Comment (jfim) [inline pinot-core/src/main/java/com/linkedin/pinot/core/query/scheduler/QueryScheduler.java]: This is likely to be too many threads, considering we shouldn't be IO bound here; since we'll end up with significantly more threads that are in executable state than the number of physical cores, they'll end up competing for execution slots/CPU cache with an overall lower throughput. There's already overcommitting of actual physical cores through SMT, so adding additional overcommitting will only lower overall throughput.

Also, this should also clarify what is the difference between a query runner and query worker, from a cursory read it's not clear which thread pool does what.

File: pinot-core/src/main/java/com/linkedin/pinot/core/query/scheduler/FCFSQueryScheduler.java (hunk 4 above)
Comment (antumbde) [inline pinot-core/src/main/java/com/linkedin/pinot/core/query/scheduler/FCFSQueryScheduler.java]: done

File: pinot-server/src/main/java/com/linkedin/pinot/server/request/ScheduledRequestHandler.java (hunk 5 above)
Comment (antumbde) [inline pinot-server/src/main/java/com/linkedin/pinot/server/request/ScheduledRequestHandler.java]: done

File: pinot-server/src/main/java/com/linkedin/pinot/server/request/ScheduledRequestHandler.java (hunk 6 above)
Comment (antumbde) [inline pinot-server/src/main/java/com/linkedin/pinot/server/request/ScheduledRequestHandler.java]: done

File: pinot-core/src/main/java/com/linkedin/pinot/core/query/scheduler/QueryScheduler.java (hunk 9 above)
Comment (antumbde) [inline pinot-core/src/main/java/com/linkedin/pinot/core/query/scheduler/QueryScheduler.java]: How and how many threads we use is a TODO (See comment). We are I/O bound on some clusters.  And it's number of running threads that will determine the cost. Using less threads can also add to the wait time. In short...these are arbitrary to give us a start.

Added code comment to explain the purpose of two pools

File: pinot-transport/src/main/java/com/linkedin/pinot/transport/netty/NettyServer.java (hunk 8 above)
Comment (antumbde) [inline pinot-transport/src/main/java/com/linkedin/pinot/transport/netty/NettyServer.java]: done

File: pinot-transport/src/main/java/com/linkedin/pinot/transport/netty/NettyServer.java [hunk 10]
```diff
@@ -193,69 +191,74 @@ public NettyChannelInboundHandler(RequestHandler handler, NettyServerMetrics met
...
-      ByteBuf responseBuf = Unpooled.wrappedBuffer(response);
-      _lastSendResponseLatency = MetricsHelper.startTimer();
-      ChannelFuture f = ctx.writeAndFlush(responseBuf);
-      _state = State.RESPONSE_WRITTEN;
-      f.addListener(this);
-      request.release();
+      final TimerContext requestProcessingLatency = MetricsHelper.startTimer();
+      final ChannelHandlerContext requestChannelHandlerContext = ctx;
+      ListenableFuture<byte[]> serializedQueryResponse = _handler.processRequest(ctx, request);
+      Futures.addCallback(serializedQueryResponse, new FutureCallback<byte[]>() {
```
Comment (kishoreg) [inline pinot-transport/src/main/java/com/linkedin/pinot/transport/netty/NettyServer.java]: when will the FutureCallback get invoked? is it internally put into another  queue and can be executed in any order. 
My main concern is we have already processed the request and response is ready but instead of sending it right away we put it in another executor service to send the response back. 

I haven't looked at the internals of the implementation but java docs say "This overload, which does not accept an executor, uses directExecutor, a dangerous choice in some cases. See the discussion in the ListenableFuture.addListener documentation.".
Looks like directExecutor means execute runnable in the same thread. Why do they say it can be dangerous?

Also, this code can benefit with some formatting. It's very hard to parse especially with the anonymous inner classes in the arguments.

File: pinot-core/src/main/java/com/linkedin/pinot/core/query/scheduler/QueryScheduler.java [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+  //
+  protected ListeningExecutorService queryRunners;
+
+  // TODO: in future, this should be driven by configured policy
+  // like poolPerTable or QoS based pools etc
+  // The policy should also determine how many threads we use per query
+
+  // These are worker threads to parallelize execution of query operators
+  // across groups of segments.
+  protected ListeningExecutorService queryWorkers;
```
Comment (kishoreg) [inline pinot-core/src/main/java/com/linkedin/pinot/core/query/scheduler/QueryScheduler.java]: where is this used?

File: pinot-core/src/main/java/com/linkedin/pinot/core/query/scheduler/QueryScheduler.java (hunk 9 above)
Comment (kishoreg) [inline pinot-core/src/main/java/com/linkedin/pinot/core/query/scheduler/QueryScheduler.java]: how do we come up with the configuration that will keep the same behavior as existing code?. Shouldn't we use cachedthreadpool to maintain existing behavior.

**Comment by kishoreg** [comment]
How did we verify that the behavior is same as existing code. We should probably use directExecutor in FCFSScheduler to maintain existing behavior.

File: pinot-transport/src/main/java/com/linkedin/pinot/transport/netty/NettyServer.java (hunk 10 above)
Comment (antumbde) [inline pinot-transport/src/main/java/com/linkedin/pinot/transport/netty/NettyServer.java]: It will be immediately executed on the same thread. Executing on the same thread can be dangerous, say, for UI threads or "main" server thread which can make the system non-responsive. In our case, it will be executed on the queryRunner which is what we want

File: pinot-core/src/main/java/com/linkedin/pinot/core/query/scheduler/QueryScheduler.java (hunk 11 above)
Comment (antumbde) [inline pinot-core/src/main/java/com/linkedin/pinot/core/query/scheduler/QueryScheduler.java]: Not used yet. This is intended to be replacement for per table executor pool going forward

File: pinot-core/src/main/java/com/linkedin/pinot/core/query/scheduler/QueryScheduler.java (hunk 9 above)
Comment (antumbde) [inline pinot-core/src/main/java/com/linkedin/pinot/core/query/scheduler/QueryScheduler.java]: This PR maintains existing code by keeping all this code unused :)
The PR commit comment explains this above - I added ScheduledRequestHandler but we continue to use existing SimpleRequestHandler. (See SimpleRequestHandlerFactory) So, the only change impacting change here is to use Futures to return from SimpleRequestHandler.  QueryExecution will continue to used the existing cached thread pool from TableDataManager.

Comment (antumbde) [inline pinot-core/src/main/java/com/linkedin/pinot/core/query/scheduler/QueryScheduler.java]: Regarding number of QueryRunner threads, we can configure it. Today, we used fixed size of 20 threads. I can change this to be similar to match that.

**Comment by antumbde** [comment]
This PR maintains existing behavior by:
1. Continuing to use SimpleRequeststHandler returned by SimpleRequestHandlerFactory. We do not use newly added ScheduledRequestHandler *yet*
2. Only impacting change her is to return Future from SimpleRequestHandler but if you see code, we execute the code as earlier and call Future.immediate() to set the future value. (IOW, we return Future only after its value is available so it's not really "future"). The future listener will execute in the same thread so nothing changes here.
I've done tests locally and on perf cluster, including perf test, for this code change

Regarding FCFSScheduler:
1. First and foremost, it's unused for now
2. I've tested it on our perf cluster for firing our existing performance workload (single query at a time) and also at higher throughputs. There is not change in performance.

Going forward:
1. We will be able to use ScheduledRequestHandler or SimpleRequestHandler based on configuration (rollback capability)
2. We *may* replace the cached thread pool in table data manager if peformance results indicate that.

**Comment by kishoreg** [comment]
comment on FCFSScheduler says that its same as existing logic but that's not true. Is it possible to split this RB into two. Make the changes to interfaces and have an implementation that mimics the existing behavior. Once that is done bring in the new implementations. For example, you can have an implementation which simply invokes the query immediately.

**Comment by antumbde** [comment]
@kishoreg FCFS is scheduling the query immediately. Why do you think logic is different ?
The new implementations are unused but the QueryScheduler is required for interface changes.

File: pinot-transport/src/main/java/com/linkedin/pinot/transport/netty/NettyServer.java [hunk 12]
```diff
@@ -193,69 +191,74 @@ public NettyChannelInboundHandler(RequestHandler handler, NettyServerMetrics met
...
+        public void onSuccess(@Nullable byte[] result) {
+          if (result == null) {
+            result = new byte[0];
+          }
+          sendResponse(result);
+        }
+
+        @Override
+        public void onFailure(Throwable t) {
+          LOGGER.error("Request processing returned unhandled exception, error: ", t);
```
Comment (mcvsubbu) [inline pinot-transport/src/main/java/com/linkedin/pinot/transport/netty/NettyServer.java]: Can you please add the query ID (and query) here? I assume that we will not get the other log what has the latency, etc.

Comment (antumbde) [inline pinot-transport/src/main/java/com/linkedin/pinot/transport/netty/NettyServer.java]: we don't have query id in this context.
Also, exception here will almost certainly not about query. It will be more because executor threw an exception.

File: pinot-server/src/main/java/com/linkedin/pinot/server/request/ScheduledRequestHandler.java [hunk 13]
```diff
@@ -0,0 +1,168 @@
...
+    this.serverMetrics = serverMetrics;
+  }
+
+  @Override
+  public ListenableFuture<byte[]> processRequest(ChannelHandlerContext channelHandlerContext,
+      ByteBuf request) {
+    final long queryStartTime = System.nanoTime();
+    serverMetrics.addMeteredGlobalValue(ServerMeter.QUERIES, 1);
+
+    LOGGER.info("Processing request : {}", request);
```
Comment (antumbde) [inline pinot-server/src/main/java/com/linkedin/pinot/server/request/ScheduledRequestHandler.java]: Fixed. Sorry, this was local change that made to PR

File: pinot-server/src/main/java/com/linkedin/pinot/server/request/ScheduledRequestHandler.java [hunk 14]
```diff
@@ -0,0 +1,168 @@
...
+
+    // transform the DataTable to serialized byte[] to send back to broker
+    ListenableFuture<byte[]> serializedQueryResponse = Futures.transform(queryResponse, new Function<DataTable, byte[]>() {
+      @Nullable
+      @Override
+      public byte[] apply(@Nullable DataTable instanceResponse) {
+        long totalNanos = System.nanoTime() - schedulerSubmitTime;
+        long serializationStartTime = System.nanoTime();
+        serverMetrics.addPhaseTiming(brokerRequest, ServerQueryPhase.QUERY_PROCESSING, totalNanos);
+        if (instanceResponse == null) {
```
Comment (antumbde) [inline pinot-server/src/main/java/com/linkedin/pinot/server/request/ScheduledRequestHandler.java]: yes. updated. That was the goal of making common method

File: pinot-server/src/main/java/com/linkedin/pinot/server/request/ScheduledRequestHandler.java [hunk 15]
```diff
@@ -0,0 +1,168 @@
...
+          LOGGER.error("Got exception while serializing response for requestId: {}, brokerId: {}",
+              instanceRequest.getRequestId(), instanceRequest.getBrokerId(), e);
+          responseBytes = null;
+        }
+        long serializationEndTime = System.nanoTime();
+        serverMetrics.addPhaseTiming(brokerRequest, ServerQueryPhase.RESPONSE_SERIALIZATION,
+            serializationEndTime - serializationStartTime);
+        serverMetrics.addPhaseTiming(brokerRequest, ServerQueryPhase.TOTAL_QUERY_TIME,
+            serializationEndTime - queryStartTime);
+
```
Comment (antumbde) [inline pinot-server/src/main/java/com/linkedin/pinot/server/request/ScheduledRequestHandler.java]: that will continue to be logged by query executor (not changed by this PR)
